### PR TITLE
Ensure stability of QGIS FeatureId when reloading layer [WFS provider] 

### PR DESCRIPTION
**Pull request by rouault**
Backport to 3.4 of https://github.com/qgis/QGIS/pull/9105

Fix http://issues.qgis.org/issues/20865

### CODE REVIEW

File: src/providers/wfs/qgswfsshareddata.cpp [hunk 1]
```diff
@@ -466,6 +477,35 @@ bool QgsWFSSharedData::createCache()
...
+    // gmlid is the gmlid or fid attribute coming from the GML GetFeature response
+    // qgisId is the feature id of the features returned to QGIS. That one should remain the same for a given gmlid even after a layer reload
+    // dbId is the feature id of the Spatialite feature in mCacheDataProvider. It might change for a given gmlid after a layer reload
+    ok &= mCacheIdDb.exec( QStringLiteral( "CREATE TABLE id_cache(gmlid TEXT, dbId INTEGER, qgisId INTEGER)" ), errorMsg ) == SQLITE_OK;
+    ok &= mCacheIdDb.exec( QStringLiteral( "CREATE INDEX idx_gmlid ON id_cache(gmlid)" ), errorMsg ) == SQLITE_OK;
+    ok &= mCacheIdDb.exec( QStringLiteral( "CREATE INDEX idx_dbId ON id_cache(dbId)" ), errorMsg ) == SQLITE_OK;
+    ok &= mCacheIdDb.exec( QStringLiteral( "CREATE INDEX idx_qgisId ON id_cache(qgisId)" ), errorMsg ) == SQLITE_OK;
+    if ( !ok )
+    {
+      QgsDebugMsg( errorMsg );
```
Comment (m-kuhn) [inline src/providers/wfs/qgswfsshareddata.cpp]: I wonder if this should be more prominent (and logged to the message panel instead)

File: src/providers/wfs/qgswfsfeatureiterator.cpp [hunk 2]
```diff
@@ -1238,6 +1249,19 @@ bool QgsWFSFeatureIterator::fetchFeature( QgsFeature &f )
 
     copyFeature( cachedFeature, f );
     geometryToDestinationCrs( f, mTransform );
+
+    // Retrieve the user-visible id from the Spatialite cache database Id
```
Comment (m-kuhn) [inline src/providers/wfs/qgswfsfeatureiterator.cpp]: Spatialite or sqlite?

File: src/providers/wfs/qgswfsshareddata.cpp (hunk 1 above)
Comment (rouault) [reply src/providers/wfs/qgswfsshareddata.cpp]: I don't anticipate this error to happen in practice. debug is probably fine for now

File: src/providers/wfs/qgswfsfeatureiterator.cpp (hunk 2 above)
Comment (rouault) [reply src/providers/wfs/qgswfsfeatureiterator.cpp]: Spatialite. There are 2 caches: a Spatialite one with the feature with its geom and attributes that gets recreated when reloadData() is called, and the new one SQlite only that only caches gmlid, qgisId and dbId.

**Review by m-kuhn — APPROVED** [review]

**Pull request merged**